File: src/com/easy/SecondMinimumNodeInABinaryTree_671.java
/**
 * 
 */
package com.easy;

import java.util.Deque;
import java.util.LinkedList;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class SecondMinimumNodeInABinaryTree_671 {

    public int findSecondMinimumValue(TreeNode root) {
        if (root == null) {
            return -1;
        }
        // return BFS(root);
        return DFS(root, root.val);
    }

    // Use DFS:
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 34.2 MB, less than 100.00%
    private int DFS(TreeNode root, int mMin) {
        if (root == null) {
            return -1;
        }

        // if root's value greater than mMin,
        // then according to tree's characteristic all children's value should
        // be >= mMin
        // So no need to visit children
        if (root.val > mMin) {
            return root.val;
        }

        int mMinLeft = DFS(root.left, mMin);
        int mMinRight = DFS(root.right, mMin);

        if (mMinLeft == -1) {
            return mMinRight;
        }
        if (mMinRight == -1) {
            return mMinLeft;
        }

        return Math.min(mMinLeft, mMinRight);
    }

    // Use BFS:
    // Time Complexity: O(n)
    // Space Complexity: O(n)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 33.9 MB, less than 100.00%
    private int BFS(TreeNode root) {
        if (root == null) {
            return -1;
        }

        int mMin = root.val;
        int mSecondMin = Integer.MAX_VALUE;
        boolean bIsFound = false;

        Deque<TreeNode> mDeQueue = new LinkedList<>();
        mDeQueue.push(root);

        while (!mDeQueue.isEmpty()) {
            TreeNode mNode = mDeQueue.pop();

            // consider value may equal integer max
            if (mNode.val > mMin && mNode.val <= mSecondMin) {
                mSecondMin = mNode.val;
                bIsFound = true;
                continue;
            }

            if (mNode.left == null) {
                continue;
            }
            mDeQueue.push(mNode.left);
            mDeQueue.push(mNode.right);
        }

        return bIsFound ? mSecondMin : -1;
    }
}

File: Golang_src/easy/secondMinimumNodeInABinaryTree_671.go
package easy

import "sort"

// Use DFS:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 1.9 MB, less than 20.00%
func findSecondMinimumValue(root *TreeNode) int {
	return dfs2(root, root.Val)
}

func dfs2(root *TreeNode, target int) int {
	if root == nil {
		return -1
	}

	if root.Val > target {
		return root.Val
	}

	left := dfs2(root.Left, target)
	right := dfs2(root.Right, target)

	if left == -1 {
		return right
	}
	if right == -1 {
		return left
	}

	return min(left, right)
}

// Use inorder traversal + sort:
// Time Complexity: O(n+nlogn)
// Space Complexity:O(n)
// Runtime: 2 ms, faster than 18.18%
// Memory Usage: 2 MB, less than 63.64%
func findSecondMinimumValue2(root *TreeNode) int {
	arr := inorder(root)
	sort.Ints(arr)

	var inorder func(root *TreeNode) []int
	inorder = func(root *TreeNode) []int {
		result := []int{}
		if root == nil {
			return result
		}
		result = append(result, inorder(root.Left)...)
		result = append(result, root.Val)
		result = append(result, inorder(root.Right)...)

		return result
	}

	min := arr[0]
	for _, v := range arr {
		if v > min {
			return v
		}
	}

	return -1
}
